balancer/catalyst/catalyst_balancer.go

```diff
@@ -0,0 +1,283 @@
+package catalyst
+
+import (
+	"context"
+	"fmt"
+	"github.com/livepeer/catalyst-api/cluster"
+	"github.com/livepeer/catalyst-api/log"
+	"math/rand"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+)
+
+type CataBalancer struct {
+	Nodes         map[string]*Node
+	nodesLock     sync.Mutex
+	Streams       map[string]Streams     // Node name -> Streams
+	IngestStreams map[string]Streams     // Node name -> Streams
+	NodeMetrics   map[string]NodeMetrics // Node name -> NodeMetrics
```

Comment:
https://pkg.go.dev/sync#Map - might be better for thread safety with its functions. Just sth to be considered in future.